plugin/commands.go

```diff
@@ -96,6 +115,100 @@ func parseInstallTarget(target string) (*installTarget, error) {
 	return it, nil
 }
 
+// Create a directory for plugin install
+func setupPluginDir(prefix string) (string, error) {
+	if prefix == "" {
+		prefix = "/opt/mackerel-agent/plugins"
+	}
+	err := os.MkdirAll(filepath.Join(prefix, "bin"), 0755)
+	if err != nil {
+		return "", err
+	}
+	return prefix, nil
+}
+
+// Download plugin artifact from `url` to `workdir`,
+// and returns downloaded filepath
+func downloadPluginArtifact(url, workdir string) (fpath string, err error) {
+	logger.Log("", fmt.Sprintf("Downloading %s", url))
+
+	// Create request to download
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("User-Agent", "mkr-plugin-installer/0.0.0")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		err = fmt.Errorf("http response not OK. code: %d, url: %s", resp.StatusCode, url)
+		return "", err
+	}
+
+	// fpath is filepath where artifact will be saved
+	fpath = filepath.Join(workdir, path.Base(url))
+
+	// download artifact
+	file, err := os.OpenFile(fpath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	_, err = io.Copy(file, resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return fpath, nil
+}
+
+// Extract artifact and install plugin
+func installByArtifact(artifactFile, bindir, workdir string) error {
+	// unzip artifact to work directory
+	err := archiver.Zip.Open(artifactFile, workdir)
+	if err != nil {
+		return err
+	}
+
+	// Look for plugin files recursively, and place those to binPath
+	return filepath.Walk(workdir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+
+		// a plugin file should be executable, and have specified name.
+		name := info.Name()
+		isExecutable := (info.Mode() & 0111) != 0
+		if isExecutable && looksLikePlugin(name) {
+			return placePlugin(path, filepath.Join(bindir, name))
+		}
+
+		// `path` is a file but not plugin.
+		return nil
+	})
+}
+
+func looksLikePlugin(name string) bool {
+	return strings.HasPrefix(name, "check-") || strings.HasPrefix(name, "mackerel-plugin-")
+}
+
+func placePlugin(src, dest string) error {
+	_, err := os.Stat(dest)
+	if err == nil {
+		logger.Log("", fmt.Sprintf("%s already exists. Skip installing for now", dest))
```

Comment:
With this implementation we cannot upgrade existing plugin. What do you think?

Reply:
I am going to implement `--force` option to override a plugin forcefully.  this can be used for upgrade also.